shared-module/displayio/Bitmap.c

```diff
@@ -79,6 +79,13 @@ void common_hal_displayio_bitmap_construct_from_buffer(displayio_bitmap_t *self,
     self->dirty_area.y2 = height;
 }
 
+void common_hal_displayio_bitmap_deinit(displayio_bitmap_t *self) {
+    self->data = NULL;
```

Comment:
Ideally the data buffer would be owned by the Bitmap and you would explicitly free it here. This gives the behavior that the large memory allocation is always freed, no matter how unlucky you are with the garbage collector.

The buffer *is* usually owned by the Bitmap except in one case for `espcamera` where it seems to borrowed from the camera driver.

Reply:
> Ideally the data buffer would be owned by the Bitmap and you would explicitly free it here. This gives the behavior that the large memory allocation is always freed, no matter how unlucky you are with the garbage collector.

I thought freeing the explicitly at this time but I'm not sure if that is best practice for dealing with the gc? The situation I ran into with it being on the stack is extremally rare. But it would work. @tannewt or @dhalbert ?
 
> The buffer _is_ usually owned by the Bitmap except in one case for `espcamera` where it seems to borrowed from the camera driver.

For anything that creates a Bitmap and includes its own buffer we would have to track if we allocated it or not. Not hard to do just another flag.
I

Reply:
I still don't understand if the pointer in question is in the live stack, or is in a frame from a routine that has returned. If it is the latter, then the stack gc is being sloppy and including stuff that is not in the current and previous frames. If it were made correct, then there would not be  problem, because the pointer would not be found.

Reply:
The best I can tell when `gc.collect()` is called the pointer exists in the live stack memory but was set in a previous function call that is complete. The part of the stack the pointer is stored in was never changed by future function calls when they allocated items on the stack.

The problem goes away with less optimization and does not exist on other ports. A Micropython video says that this situation, while rare, is possible. I have seen a cases where local variables exist on the stack but are not initialized and unless a particular code path runs they are never overwritten. If that occurs the original value still exists on the stack. A future call may happen to re-use this same memory space but never changes it.

Reply:
> Is this bitmap referenced by anything else? TileGrid maybe? Otherwise you could do an explicitly free here.

It's available as `.bitmap`, right?, so someone else could have a reference to it.

Reply:
> Is this bitmap referenced by anything else? TileGrid maybe? Otherwise you could do an explicitly free here.

TileGrid calls _get_pixel so it has no direct access to bitmap's data element. Added in the explicit free.

> It's available as .bitmap, right?, so someone else could have a reference to it.

This is only to the `Bitmap` object in `OnDiskGif`. For `Bitmap` the data element is not accessed by any other object. (I searching through all `shared-module` for `data` and checked).

If the data buffer was passed to the bitmap (as seems to be the case for the camera) the buffer will not be freed automatically.

Reply:
Explicitly calling `gc_free` on data is making the build over size on the few M0 boards that include displayio.

Ideas (in my general order of best idea best to worst):
1. Revert back to having the GC free data and my code setting the data pointer to NULL. This original change fit and still fixes the original issue.
2. `#ifdef SAMD21` (or similar) around the new code. I'm not sure if using port specific `#ifdef` in the common `shared-module` is a good idea.
3. Try to find something else to save 24 bytes or so.

Reply:
I will look at space savings. We need something more, because anything that is almost about to overflow is going to happen sooner or later. It _is_ a question whether we should include this on the smallest boards. Hallowing M0 is the only one I'm thinking might be worthwhile.

Reply:
is there any evidence that OnDiskGif is useful within the RAM constraints of the HalloWing M0?

Reply:
> is there any evidence that OnDiskGif is useful within the RAM constraints of the HalloWing M0?

Quickly checked, with only 32Kb RAM it would not be worthwhile if you ever managed to get it to work. Maybe something added on to go straight from GIF->Display but that would be the only way I see it possible. `OnDiskGif` is currently turned off on all M0 builds.

The overflowing part is with the changes to `displayio.Bitmap`